bip-0352.mediawiki

```diff
@@ -0,0 +1,477 @@
+
+''' Simple case '''
+
+Bob publishes a public key ''B'' as a silent payment address. Alice discovers Bob's silent payment address, selects a UTXO with private key ''a'', public key ''A'' and creates a destination output ''P'' for Bob in the following manner:
+
+* Let ''P = B + hash(a·B)·G''
+* Encode ''P'' as a [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki BIP341] taproot output
+
+Since ''a·B == b·A'' ([https://en.wikipedia.org/wiki/Elliptic-curve_Diffie%E2%80%93Hellman Elliptic-curve Diffie–Hellman]), Bob scans with his private key ''b'' by collecting the input public keys for each transaction with at least one unspent taproot output and performing the ECDH calculation until ''P'' is found (i.e. calculating ''P = B + hash(b·A)·G'' and seeing that ''P'' is present in the transaction outputs).
+
+''' Creating more than one output '''
+
+In order to allow Alice to create more than one output for Bob<ref name="why_more_than_one_output">'''Why allow for more than one output?''' Allowing Alice to break her payment to Bob into multiple amounts opens up a number of privacy improving techniques for Alice, making the transaction look like a CoinJoin or better hiding the change amount by splitting both the payment and change outputs into multiple amounts. It also allows for Alice and Carol to both have their own unique output paying Bob in the event they are in a collaborative transaction and both paying Bob's silent payment address.</ref>, we include an integer in the following manner:
+
+* Let ''k = 0''
+* Let ''P<sub>0</sub> = B + hash(a·B || n)·G''
+* For additional outputs:
+** Increment ''k'' by one (''k++'')
+** Let ''P<sub>i</sub> = B + hash(a·B || n)·G''
+
+Bob detects this output the same as before by searching for ''P<sub>0</sub> = B + hash(b·A || 0)·G''. Once he detects the first output, he must:
+
+* Check for ''P<sub>1</sub> = B + hash(b·A || 1)·G''
+* If ''P<sub>1</sub>'' is not found, stop
+* If ''P<sub>1</sub>'' is found, continue to check for ''P<sub>2</sub>'' and so on until an additional output is not found
+
+Since Bob will only perform these subsequent checks after a transaction with at least one output paying him is found, the increase to his overall scanning requirement is negligible.
+
+''' Preventing address reuse '''
+
+If Alice were to use a different UTXO from the same public key ''A'' for a subsequent payment to Bob, she would end up deriving the same destination ''P''. To prevent this, Alice should include a hash of the outpoint in the following manner:
+
+* Let ''outpoint_hash = hash(txid || vout)''
+* Let ''P<sub>0</sub> = B + hash(outpoint_hash·a·B || 0)·G''
+
+Bob must include the same ''outpoint_hash'' when scanning.
```

Comment:
Neat, that makes sense.

Reply:
Hm, I had the same question. Your answer makes sense, but I worry that there can be strange interactions of the multiplicands.

For example, with two different `outpoint_hash = outpoint_hash'`, I can create
 * `P0 = B + hash(outpoint_hash·a·B || 0)·G` and
 * `P0' = B + hash(outpoint_hash'·a'·B || 0)·G = P0` where `a' = a · outpoint_hash / output_hash'`

Is this a problem? I don't know. At least, that seems to contradict the idea that even malicious senders shouldn't be able to create collisions (without creating double spends). And the receiver can do the same thing.

If you want to avoid this but still have 33 bytes, you could do two steps of tweaking:
 -  `B' = B + hash(outpoint_hash||B) · G`
 -  `P0 = B' + hash(a' · B' || 0) · G`

(Hm, I have a suspicion that now you will tell me that this breaks something because now there's a B in the first hash...)

Reply:
@real-or-random 

Apologies for the late reply, @josibake informed me that this was still unaddressed.

Initially I wanted to say `a' = a · outpoint_hash / outpoint_hash'` has a circular reference, because `outpoint_hash'` contains the hash of `a'` and thus you can never actually calculate `a'`.

However, we're currently discussing the possibility of [hashing only one outpoint](https://github.com/bitcoin/bips/pull/1458#discussion_r1378773586) instead of all of them, and then this protective circular reference would be gone.

I think your suggested mitigation still requires 65 bytes for light clients. Light clients can't calculate `B' = B + hash(outpoint_hash||B) · G` followed by `P0 = B' + hash(b' · A' || 0) · G` without learning both `outpoint_hash` (32 bytes) and `A'` (33 bytes).

It seems to me that another way to fix this is by reintroducing the circular reference with `P0 = B + hash(hash(outpoint_hash || A)·b·A || 0)·G` (also doesn't require another ECC multiplication) . Now light clients only need to learn `hash(outpoint_hash || A)·A` (33 bytes) to calculate `P0`.

Reply:
In summary, I think [only hashing one outpoint](https://github.com/bitcoin/bips/pull/1458#discussion_r1378773586) is the right way to go, so this is something we should address.

You are correct @RubenSomsen that @real-or-random 's proposal requires 65 bytes for light clients, and I don't see any issues with your proposal of `hash(outpoint || A)·A`. The only correction would be we can concatenate the `outpoint` and `A` directly (no need to hash the outpoint first). I'm assuming this is what you meant, but just wanted to clarify. To check my understanding, this works because `A` is the sum of silent payment eligible inputs, so any time the attacker tries to maliciously choose `a'`, `A'` would be included in the hash.

@real-or-random thoughts on the updated proposal?

Reply:
also, @benma can you confirm this is feasible from a HWW perspective?

Reply:
@RubenSomsen @josibake I agree with your comments, and the changes in 0985f299b3420680cdc32f56f01391e3c5b8064f look good to me.